controllers/patcher.go

```diff
@@ -0,0 +1,141 @@
+	"sigs.k8s.io/kustomize/api/resmap"
+	kustypes "sigs.k8s.io/kustomize/api/types"
+	"sigs.k8s.io/kustomize/kyaml/filesys"
+
+	configv1beta1 "github.com/projectsveltos/addon-controller/api/v1beta1"
+)
+
+// Kustomize is a Helm post-render plugin that runs Kustomize.
+type CustomPatchPostRenderer struct {
+	// Patches is a list of patches to apply to the rendered manifests.
+	Patches []configv1beta1.Patch
+}
+
+func (k *CustomPatchPostRenderer) Run(renderedManifests *bytes.Buffer) (modifiedManifests *bytes.Buffer, err error) {
+	fs := filesys.MakeFsInMemory()
+	cfg := kustypes.Kustomization{}
+	cfg.APIVersion = kustypes.KustomizationVersion
+	cfg.Kind = kustypes.KustomizationKind
+
+	// Add rendered Helm output as input resource to the Kustomization.
+	const input = "helm-output.yaml"
+	cfg.Resources = append(cfg.Resources, input)
+	if err := writeFile(fs, input, renderedManifests); err != nil {
+		return nil, err
+	}
+
+	// Add patches.
+	for _, m := range k.Patches {
+		cfg.Patches = append(cfg.Patches, kustypes.Patch{
+			Patch:  m.Patch,
+			Target: adaptSelector(m.Target),
+		})
+	}
+
+	// Write kustomization config to file.
+	kustomization, err := json.Marshal(cfg)
+	if err != nil {
+		return nil, err
+	}
+	if err := writeToFile(fs, "kustomization.yaml", kustomization); err != nil {
+		return nil, err
+	}
+	resMap, err := buildKustomization(fs, ".")
+	if err != nil {
+		return nil, err
+	}
+	yaml, err := resMap.AsYaml()
+	if err != nil {
+		return nil, err
+	}
+	return bytes.NewBuffer(yaml), nil
+}
+
+func writeToFile(fs filesys.FileSystem, path string, content []byte) error {
+	helmOutput, err := fs.Create(path)
+	if err != nil {
+		return err
+	}
+	if _, err = helmOutput.Write(content); err != nil {
+		return err
+	}
+	if err = helmOutput.Close(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func writeFile(fs filesys.FileSystem, path string, content *bytes.Buffer) error {
+	helmOutput, err := fs.Create(path)
+	if err != nil {
+		return err
+	}
+	if _, err = content.WriteTo(helmOutput); err != nil {
+		return err
+	}
+	if err = helmOutput.Close(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func adaptSelector(selector *configv1beta1.PatchSelector) (output *kustypes.Selector) {
+	if selector != nil {
+		output = &kustypes.Selector{}
+		output.Gvk.Group = selector.Group
+		output.Gvk.Kind = selector.Kind
+		output.Gvk.Version = selector.Version
+		output.Name = selector.Name
+		output.Namespace = selector.Namespace
+		output.LabelSelector = selector.LabelSelector
+		output.AnnotationSelector = selector.AnnotationSelector
+	}
+	return
+}
+
+// TODO: remove mutex when kustomize fixes the concurrent map read/write panic
+var kustomizeRenderMutex sync.Mutex
+
+// buildKustomization wraps krusty.MakeKustomizer with the following settings:
+// - load files from outside the kustomization.yaml root
+// - disable plugins except for the builtin ones
+func buildKustomization(fs filesys.FileSystem, dirPath string) (resmap.ResMap, error) {
```

Comment:
Thanks. We can this function also from [here](https://github.com/projectsveltos/addon-controller/blob/4d8403710b08cfc2d5fb66f25df7f17dd3f8b057/controllers/handlers_kustomize.go#L485)